src/concurrent/OneToOneRingBuffer.h

```diff
@@ -0,0 +1,236 @@
+        buffer_(buffer),
+        capacity_(length - RingBufferDescriptor::TRAILER_LENGTH),
+        descriptor_(reinterpret_cast<RingBufferDescriptor::RingBufferDescriptorDefn *>(buffer_ + capacity_)),
+        tailPosition_(buffer_ + capacity_ + offsetof(RingBufferDescriptor::RingBufferDescriptorDefn, tailPosition)),
+        headCachePosition_(buffer_ + capacity_ + offsetof(RingBufferDescriptor::RingBufferDescriptorDefn, headCachePosition)),
+        headPosition_(buffer_ + capacity_ + offsetof(RingBufferDescriptor::RingBufferDescriptorDefn, headPosition)),
+        correlationCounter_(buffer_ + capacity_ + offsetof(RingBufferDescriptor::RingBufferDescriptorDefn, correlationCounter)),
+        maxMsgLength_(capacity_ / 8)
+    {
+        checkCapacity(capacity_);
+    }
+
+    OneToOneRingBuffer(const OneToOneRingBuffer &) = delete;
+    OneToOneRingBuffer& operator=(const OneToOneRingBuffer &) = delete;
+
+    inline uint32_t capacity() const noexcept
+    {
+        return capacity_;
+    }
+
+    inline uint32_t maxMsgLength() const noexcept
+    {
+        return maxMsgLength_;
+    }
+
+    position_t write(std::int32_t msgTypeId, const uint8_t *srcBuffer, uint32_t length) noexcept
+    {
+        if (msgTypeId < 1)
+        {
+            return RingBufferDescriptor::INVALID_ARGUMENT;
+        }
+
+        if (length > maxMsgLength_)
+        {
+            return RingBufferDescriptor::INVALID_ARGUMENT;
+        }
+
+        const uint32_t recordLength = length + RingBufferDescriptor::HEADER_LENGTH;
+        const uint32_t requiredCapacity = RingBufferDescriptor::align(recordLength, RingBufferDescriptor::ALIGNMENT);
+        const uint32_t mask = capacity_ - 1;
+
+        std::uint64_t head = descriptor_->headCachePosition;
+        std::uint64_t tail = descriptor_->tailPosition;
+        const size_t availableCapacity = capacity_ - (tail - head);
+
+        if (requiredCapacity > availableCapacity)
+        {
+            head = getUInt64Volatile(headPosition_);
+
+            if (requiredCapacity > (capacity_ - (tail - head)))
+            {
+                return RingBufferDescriptor::INSUFFICIENT_SPACE;
+            }
+
+            descriptor_->headCachePosition = head;
+        }
+
+        uint32_t padding = 0;
+        uint32_t recordIndex = static_cast<uint32_t>(tail) & mask;
+        const uint32_t toBufferEndLength = capacity_ - recordIndex;
+
+        if (requiredCapacity > toBufferEndLength)
+        {
+            uint32_t headIndex = static_cast<uint32_t>(head) & mask;
+
+            if (requiredCapacity > headIndex)
+            {
+                head = getUInt64Volatile(headPosition_);
+                headIndex = static_cast<uint32_t>(head) & mask;
+
+                if (requiredCapacity > headIndex)
+                {
+                    return RingBufferDescriptor::INSUFFICIENT_SPACE;
+                }
+
+                putUInt64Ordered(headCachePosition_, head);
```

Comment:
not being a big expert on memory models...
do we need to use these specialized read/write methods instead of using std::atomic?

And from a higher level perspective. The whole RS-cpp implementation is written in the model that there is one thread executing the code. We chose single threaded model. With that regards, how much simpler would the OneToOneRingBuffer class look like if we don't have to deal with concurrency? Would you agree that the complexity associated with concurrency would not be used?